runtime/runtime_plugins.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2018 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+// +build linux,cgo darwin,cgo
+
+package runtime
+
+// Contains parts of the runtime package that use the plugin package
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"plugin"
+
+	"github.com/pkg/errors"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/loader"
+	"github.com/open-policy-agent/opa/plugins"
+	"github.com/open-policy-agent/opa/plugins/logs"
+	"github.com/open-policy-agent/opa/server"
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/storage/inmem"
+	"github.com/open-policy-agent/opa/topdown"
+)
+
+// NewRuntime returns a new Runtime object initialized with params.
+func NewRuntime(ctx context.Context, params Params) (*Runtime, error) {
+
+	if params.ID == "" {
+		var err error
+		params.ID, err = generateInstanceID()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	loaded, err := loader.Filtered(params.Paths, params.Filter)
+	if err != nil {
+		return nil, err
+	}
+
+	store := inmem.New()
+
+	txn, err := store.NewTransaction(ctx, storage.WriteParams)
+	if err != nil {
+		return nil, err
+	}
+
+	// only register custom plugins if directory specified
+	if params.BuiltinDir != "" {
```

Comment:
I think it ought to be possible to factor this check and the call below into a separate file that's forked to handle cross-platform support. There should be a single NewRuntime implementation that calls `err := RegisterBuiltinsForParams(params)` (or something similar) and `err := RegisterPluginsForParams(params)` below. Then there will be separate files that implement those functions (one for platforms that support plugins and one for those that do not.)

Reply:
Sure, sounds good.